processor/metricstarttimeprocessor/internal/datapointstorage/timeseries_map.go

```diff
@@ -0,0 +1,103 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package datapointstorage // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/metricstarttimeprocessor/internal/datapointstorage"
+
+import (
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+// TimeseriesInfo contains the information necessary to adjust from the initial point and to detect resets.
+type TimeseriesInfo struct {
+	Mark bool
+
+	Number               pmetric.NumberDataPoint
+	Histogram            pmetric.HistogramDataPoint
+	ExponentialHistogram pmetric.ExponentialHistogramDataPoint
+	Summary              pmetric.SummaryDataPoint
+}
+
+type TimeseriesKey struct {
+	Name           string
+	Attributes     [16]byte
+	AggTemporality pmetric.AggregationTemporality
+}
+
+// TimeseriesMap maps from a timeseries instance (metric * label values) to the timeseries info for
+// the instance.
+type TimeseriesMap struct {
+	sync.RWMutex
+	// The mutex is used to protect access to the member fields. It is acquired for the entirety of
+	// AdjustMetricSlice() and also acquired by gc().
+
+	Mark   bool
+	TsiMap map[TimeseriesKey]*TimeseriesInfo
+}
+
+// Get the TimeseriesInfo for the timeseries associated with the metric and label values.
+func (tsm *TimeseriesMap) Get(metric pmetric.Metric, kv pcommon.Map) (*TimeseriesInfo, bool) {
+	// This should only be invoked be functions called (directly or indirectly) by AdjustMetricSlice().
+	// The lock protecting tsm.tsiMap is acquired there.
+	name := metric.Name()
+	key := TimeseriesKey{
+		Name:       name,
+		Attributes: getAttributesSignature(kv),
+	}
+	switch metric.Type() {
+	case pmetric.MetricTypeHistogram:
+		// There are 2 types of Histograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.Histogram().AggregationTemporality()
+	case pmetric.MetricTypeExponentialHistogram:
+		// There are 2 types of ExponentialHistograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.ExponentialHistogram().AggregationTemporality()
+	}
+
+	tsm.Mark = true
+	tsi, ok := tsm.TsiMap[key]
+	if !ok {
+		tsi = &TimeseriesInfo{}
+		tsm.TsiMap[key] = tsi
+	}
+	tsi.Mark = true
+	return tsi, ok
+}
+
+// Create a unique string signature for attributes values sorted by attribute keys.
+func getAttributesSignature(m pcommon.Map) [16]byte {
+	clearedMap := pcommon.NewMap()
+	m.Range(func(k string, attrValue pcommon.Value) bool {
```

Comment:
If I'm reading right, this seems like a hot-path for performance and needing to duplicate the entire array with empty-valued attributes removed seems at first glance like a waste of cycles but I don't know the edge case it's handling. Maybe a comment to explain this so it's clear for the next person who looks? If there's no way around it that is.

Reply:
See the `TestEmptyLabels` test for an example of why. basically prometheus treats empty labels the same way as non-existing ones - and so they shared the same cache entry here too.

Given this now lives out of the prom receiver, we should maybe find a better path forward. I created an issue to investigate and added it as a comment here.

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38621